File: swift/Solutions/Day.swift
//
//  Day.swift
//  advent_of_code_2023
//
//  Created by Kushantha Attanayake on 11/29/23.
//

import Foundation

protocol Day {
    /// Solves part 1 of the problem
    func part1Solution(input: [String]) -> String
    
    /// Solves part 2 of the problem
    func part2Solution(input: [String]) -> String
}

extension Day {
    
    /// Validates that the implemented [part1Solution] correctly solves the example input to the problem
    func validatePart1() {
        let expectedOutput = readFile(from: File.examplePart1Out)[0]
        let actualOutput = part1Solution(input: readFile(from: File.examplePart1In))
        print("Validating part 1 ... ", terminator: "")
        assert(expectedOutput == actualOutput)
        print("Validation complete.")
    }
    
    func part1() {
        print("Solving part 1: ", terminator: "")
        print(part1Solution(input: readFile(from: File.part1)))
    }
    
    /// Validates that the implemented [part1Solution] correctly solves the example input to the problem
    func validatePart2() {
        let expectedOutput = readFile(from: File.examplePart2Out)[0]
        let actualOutput = part2Solution(input: readFile(from: File.examplePart2In))
        print("Validating part 2 ... ", terminator: "")
        assert(expectedOutput == actualOutput)
        print("Validation complete.")
    }
    
    func part2() {
        print("Solving part 2: ", terminator: "")
        print(part2Solution(input: readFile(from: File.part2)))
    }
}

// MARK: File IO extensions
extension Day {
    private var baseResourcePath: String {
        get {
            return #file
                .replacing(#"/Solutions/"#, with: "/Resources/")
                .replacing("Day.swift", with: name)
        }
    }
    
    private var name: String {
        get { return String(String(describing: self).split(separator: "(")[0]) }
    }
    
    private func readFile(from:File) -> [String] {
        do {
            return try String(contentsOfFile: baseResourcePath + from.path).split(separator: "\n").map { String($0) }
        } catch {
            print(error.localizedDescription)
            exit(1)
        }
    }
}

private enum File {
    case examplePart1In
    case examplePart1Out
    case examplePart2In
    case examplePart2Out
    case part1
    case part2
    
    var path: String {
        switch(self){
        case .examplePart1In:
            return "/ExamplePart1In.txt"
        case .examplePart1Out:
            return "/ExamplePart1Out.txt"
        case .examplePart2In:
            return "/ExamplePart2In.txt"
        case .examplePart2Out:
            return "/ExamplePart2Out.txt"
        case .part1:
            return "/Part1In.txt"
        case .part2:
            return "/Part2In.txt"
        }
    }
}

File: kotlin/src/main/kotlin/solutions/Day.kt
package solutions

import kotlin.io.path.Path
import kotlin.io.path.readLines

interface Day {

    /**
     * Solution to part 1 of a problem.
     */
    fun part1Solution(input: List<String>): String

    /**
     * Solution to part 2 of a problem
     */
    fun part2Solution(input: List<String>): String

    /**
     *  Validates that [part1Solution] correctly solves the problem example
     */
    fun validatePart1() {
        print("Validating part 1 ... ")
        check(loadFile(EXAMPLE_PART_1_ANSWER).first() == part1Solution(loadFile(EXAMPLE_PART_1_INPUT)))
        println("Validation complete")
    }

    /**
     *  Validates that [part2Solution] correctly solves the problem example
     */
    fun validatePart2() {
        print("Validating part 2 ... ")
        check(loadFile(EXAMPLE_PART_2_ANSWER).first() == part2Solution(loadFile(EXAMPLE_PART_2_INPUT)))
        println("Validation complete")
    }


    /**
     * Solves part 1 of the challenge
     */
    fun part1() {
        print("Solving part 1:")
        println(part1Solution(loadFile(PART_ONE_INPUT)))
    }

    /**
     * Solves part 1 of the challenge
     */
    fun part2() {
        print("Solving part 2:")
        println(part2Solution(loadFile(PART_TWO_INPUT)))
    }

    private fun loadFile(file: String): List<String>{
        val relativePath = this.javaClass.simpleName.lowercase()
        return Path(PATH_PREFIX + relativePath + file).readLines()
    }

    companion object {
        private const val PATH_PREFIX = "src/main/resources/"
        private const val EXAMPLE_PART_1_INPUT = "/ExamplePart1In.txt"
        private const val EXAMPLE_PART_1_ANSWER = "/ExamplePart1Out.txt"
        private const val EXAMPLE_PART_2_INPUT = "/ExamplePart2In.txt"
        private const val EXAMPLE_PART_2_ANSWER = "/ExamplePart2Out.txt"
        private const val PART_ONE_INPUT = "/Part1In.txt"
        private const val PART_TWO_INPUT = "/Part2In.txt"
    }

}